### DISCUSSION TITLE
Running Vaultwarden in a server cluster / High Availability

### DISCUSSION THREAD

**Discussion by ThomasLeister** [dani-garcia/vaultwarden#2691] [Q&A] [answered]
I've recently started using my own Vaultwarden based server. Now that part of my credentials reside on my server, I'm thinking about making it more resilient to outages. I'm aware that Bitwarden clients usually store a cached version of my password safe locally, but anyway I'd like to make Vaultwarden highly available somehow. 

The most obvious choice would be to run Vaultwarden on at least three different servers and link its MySQL databases to a cluster, e.g. via Mariadb Galera Cluster. I'm just not sure if this is enough and if sessions can easily be handled by multiple servers at the same time. If there something like an internal state of Vaultwarden that is not stored in MySQL / database and needs to be synced separately? Is Vaultwarden even capable of being run in multiple instances for one password safe?

**Comment by BlackDex** [comment, marked as answer]
I suggest to read: https://vaultwarden.discourse.group/t/is-the-bitwarden-rs-service-stateless-if-its-using-postgres-backend/450

**Reply by archef2000** [reply]
Would you be open to allow HA with all features if I would address your points?
So your points from the discussion are:
1. Websockets will not work correctly between containers.
2. If multi running containers have different rsa keys login/sessions between the containers will fail.
3. No shared icon storage. (Unless a shared filesystem is used)
4. Attachments could be scattered (Unless a shared filesystems is used)

Number 3&4 would need to be a provided by the user.
But sessions + Websockets are a bit more difficult.
From my knowledge Websockets are used for changes of the Vault to take instant effect on all clients.
And the "login with device" feature would all need quick inter-server communication.

Redis could fulfill these requirements.
Are there any other features or limitation that I am missing?
What is the biggest datasize that would be send?
How could it be made secure?

**Reply by BlackDex** [reply]
Redis is not going to help here. It's not something in regards of sessions. It's more that two servers need to communicate with each other regarding those push/websocket messages.

Regarding the RSA Key, i was already thinking about this, and i think we can even move that to the database instead of having this on-disk.

Other items are mostly on the user-side i think.

**Reply by archef2000** [reply]
But the RSA key is not going to change so expecting the user to sync that once is not the difficult and attachments&icons already need to be synced.

But redis could be used for websocket by making a "channel" for each device and a table of devices by uuid per account and then the server that is connected with the device can answer in the channel of the other device.

And the push notifications could be handled by one of the servers that then mrks them as "send" and if it is not marked as "send" by this server the next server gets elected?

**Reply by BlackDex** [reply]
That isn't really how it works.
If anything, a MQ would be better in these cases.
Each server could have websocket connections of different clients. All our some may need to be notified via a message.
Push messages to Bitwarden need to be handled by just one server and not multiple.

The gossip protocol might be an option here to instead of MQ.
That would prevent users from adding a different service into play.

**Reply by archef2000** [reply]
What do mean with a different service into play? redis?

The push messages need to be handled by just one server that is right, but what if the server crashes and does not deliver the message a different server would need to take over the "push message" role.

So you would implement a MQ in vaultwarden itself and connect the bitwarden servers directly to each other if I understood that correctly?

**Reply by BlackDex** [reply]
No. Any push messages is only sent on a change. That change is handled by one server and that server can send a message just fine too the push servers. Doesn't matter which server sends it.

That isn't the same for the websockets. Every server which has active websocket connections need to know about a change from any active server, doesn't matter if that server has websocket connections or not which receives the update from the client. That communication need to be handled somehow internally.

MQ could be a solution, but i do not really like the need to have an extra service running for this to work, like sn MQ service. That is why i mentored the gossip protocol. That should in theory be possible to do something like this.

**Reply by archef2000** [reply]
Many services use an extra service to make HA work (like redis). And what if the servers can't reach each other and only share the db and a redis connection. (If you have mulitiple locatioons that are only reachable localy for example)

Aren't push messages also used for "login with device"?

But with gossip it is like a ring token right? But would latency not be a problem with multiple servers in the chain and also they need to reach each other.

**Reply by BlackDex** [reply]
If all servers can reach the same Redis/MQ/DB server, then they probably also can talk to each other.

Push messages are indeed used for the `Login with device`. But the server in the end sends a WebSocket message with the right data to the right connection. And that connection in an HA setup can be on any server, not necessarily on the same server which receives the validation from the client device. So, that somehow needs to be communicated between all servers, hence why i say that all websocket messages need to be handled by all servers, not just one.

Gossip kinda is a ring yes. And latency isn't that huge i think. And a small amount of latency isn't that bad i think.

Also, there are different HA definitions.
HA In the sense of really High Available, if one server goes down, the other takes it over, that can be either hot or cold standby, doesn't matter that much.
Or
HA In the sense of load balancing, where each server is active and can handle all connections from any client no matter where they come from.

In my opinion, the challenges you are talking about are mostly tied to the Load Balancing example. For which i actually see no use case for Vaultwarden. It can handle the traffic well enough with just one instance. Maybe there are some differences in which type of Database  you use which can help a bit. But that is about it i think.

So, from my point of view, a standby HA setup is more relevant for a tool like Vaultwarden, and that doesn't create these issues at all.
The only thing which needs to be checked is the file storage that it is shared. The websocket connections will try to reconnect and that will work just fine on the other instance, and once all websocket connections are over to the other instance they all receive the correct messages.

**Reply by archef2000** [reply]
I would like to setup Vaultwarden on two servers so i can also update my vault if one of them is down but I would probably set the DNS up in round robin so it would be kind of load balancing. And with the servers reaching each other maybe you have some authentication infront of vaultwarden so you would need to create a seperate channel and I think that, that an extra service that is made for these things in the back is not too much to ask from the user and you can cover more use cases as it would be universal.

As redis is easy to setup for testing and also make it HA with a cluster see the [example of nextcloud](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html#organizations-with-clustered-setups).

As I don't know any MQ service that is low on resources and easy to implement and run.

What thinks would need to be "shared" besides the files if it would HA with LB?

**Reply by archef2000** [reply]
Would you be open to making Vautwarden HA (load balancing) if I would make a PR? With shared websocket and login requests?

**Reply by seagyn** [reply]
+1 on gossip. It's the only option that doesn't drag in another service to run, which I think was your main hesitation. I looked at Postgres LISTEN/NOTIFY (no new service, reuses the DB) but that only helps Postgres users, and Redis/MQ mean everyone runs an extra thing. Gossip is the one path that's no-extra-service and works on any backend.

The state that needs sharing is small and already goes through one spot: all the send_* methods build the payload with create_update() and call WebSocketUsers::send_update(user_id, data). So that's the only place anything changes. Local change pushes to local sockets like today and also forwards {user_ids, data} to peers; an inbound event just calls a local-only send_update() per user_id (no re-forward, or you loop). The forwarded blob is the already-serialized event plus the target user_id(s), a few hundred bytes. It's the same "something changed, resync" nudge the client already handles, and the client still pulls the actual data over HTTPS after, so no vault contents go over the channel. Best-effort delivery is fine here since clients reconcile on their sync interval anyway.

For membership I'd use proper SWIM via a crate like foca or memberlist: bootstrap off a seed list (a DNS name or a static list from config), then peer-to-peer with failure detection. Works anywhere, self-heals, and it's a library not a service.

On security, since it's a password manager: the gossip channel is authenticated by default (HMAC shared secret or mTLS, no unauthenticated mode, refuses to start without a secret) and bound to the internal network. No vault data ever goes over it, only the resync nudge, so the worst a compromised channel could do is trigger resyncs or probe which user_ids exist, and authentication closes that off.

All behind a flag (off by default), single-instance behaviour unchanged, foca as an optional cargo feature. Scope is just the existing UpdateType events; mobile push is untouched.

A few things I'd want your call on:

- foca or memberlist, or do you have a crate preference?
- HMAC or mTLS for the channel, or both?
- ok with an optional foca dependency behind a cargo feature?

If the overall shape sounds right I'll put together a proof of concept and show it fanning out across a few local instances before turning it into a PR.
